tools/ci_build/rewriter.py

```diff
@@ -0,0 +1,221 @@
+'''comment out unused ops in cpu_execution_provider.cc by a set of models'''
+
+import os
+import shutil
+import onnx
+from onnx import AttributeProto as AP
+
+#pylint: disable=no-member,too-many-locals,too-many-statements
+
+def extract_ops_from_file(file_path, referred_ops):
+    '''extract ops from csv file with format:
+        op_type,domain,opset'''
+
+    if not os.path.isfile(file_path):
+        return referred_ops
+
+    with open(file_path, 'r') as csv_to_read:
+
+        for line in csv_to_read.readlines():
+            op_type, domain, raw_opset = line.strip().split(',')
+            opset = int(raw_opset)
+
+            if op_type not in referred_ops:
+                referred_ops[op_type] = {domain:[opset]}
+
+            elif domain not in referred_ops[op_type]:
+                referred_ops[op_type][domain] = [opset]
+
+            elif opset not in referred_ops[op_type][domain]:
+                referred_ops[op_type][domain].append(opset)
+
+    return referred_ops #end of extract_ops_from_file(...)
+
+
+def extract_ops_from_model(model_path, referred_ops):
+    '''extract ops from models under model_path and return a diction'''
+
+    if not os.path.isdir(model_path):
+        return referred_ops
+
+    def map_domain(domain):
+        if domain == 'ai.onnx.ml':
+            return 'kMLDomain'
+        return 'kOnnxDomain'
+
+    def extract_ops_from_graph(graph, opsets, operators):
+        '''extract ops from graph and all subgraphs'''
+
+        for operator in graph.node:
+
+            if operator.op_type not in operators:
+                operators[operator.op_type] = {}
+
+            for opset in opsets:
+                mapped_domain = map_domain(opset.domain)
+                if mapped_domain not in operators[operator.op_type]:
+                    operators[operator.op_type][mapped_domain] = []
+                if opset.version not in operators[operator.op_type][mapped_domain]:
+                    operators[operator.op_type][mapped_domain].append(opset.version)
+
+            for attr in operator.attribute:
+
+                if attr.type == AP.GRAPH: #process subgraph
+                    extract_ops_from_graph(attr.g, opsets, operators)
+
+                elif attr.type == AP.GRAPHS: #process all subgraphs
+                    for subgraph in attr.graphs:
+                        extract_ops_from_graph(subgraph, opsets, operators)
+
+    for root, _, files in os.walk(model_path):
+        for file in files:
+            if file.endswith('.onnx'):
+                model_path = os.path.join(root, file)
+                model = onnx.load(model_path)
+                extract_ops_from_graph(model.graph, model.opset_import, referred_ops)
+
+    return referred_ops #end of extract_ops_from_model(...)
+
+
+def rewrite_cpu_provider(model_path, file_path, ep_path):
+    '''rewrite provider file to exclude unused ops'''
+
+    onnx_op = 'ONNX_OPERATOR_KERNEL_CLASS_NAME'
```

Comment:
i wonder if it's possible to accomplish the disabling of certain ops some other way? for example, by adding compile definitions like DISABLE_OP_X (though the macro stuff could be ugly...).
some concerns with this text manipulation:
- parsing corner cases and brittleness if the the constructs matched here change in the source code
- modifying source files may cause accidental changes to be commited

Reply:
@pranavsharma : comments?

Reply:
Can you create a separated cc file for the reduced ops?
For example,  cpu_execution_provider.cc, you can 
- Generate a cpu_execution_provider_reduced_ops.cc with the original  cpu_execution_provider.cc untouched,
- Set cmake options such as onnxruntime_BUILD_REDUCED_OPS  for onnxruntime_providers.cmake to exclude the original cpu_execution_provider.cc and include cpu_execution_provider_reduced_ops.cc for onnxruntime_providers_srcs?

Reply:
Doing it via #defines would equate to a #define for every forward declaration and registration in the EP definition. 

It's possible to do some template magic to remove entire opsets, but that's not fine grained enough. 

---
In reply to: [468270957](https://github.com/microsoft/onnxruntime/pull/4747#discussion_r468270957) [](ancestors = 468270957)

Reply:
Some test infrastructure to validate the disabling of ops is working correctly would be good.

If we follow the suggestion from @gwang-msft to write to filenames with 'reduced' in it we could add those filenames to the git ignore list so they're not accidentally checked in.

---
In reply to: [468917137](https://github.com/microsoft/onnxruntime/pull/4747#discussion_r468917137) [](ancestors = 468917137,468270957)

Reply:
Yes - believe we need a pipeline to package a reduced whl and test with corresponding models, agree?
For git-ignore I am having a flipped perspective - may be user want to know the file has been changed with a simple 'git status'. Further , the chance of accidental check-in is minimal since we have code reviews and commit is always reversible.

Reply:
Renamed the provider files to "...~" which is already excluded by .gitignore.